eng/setup-pr-validation.ps1

```diff
@@ -1,24 +1,40 @@
 [CmdletBinding(PositionalBinding=$false)]
 param (
   [string]$sourceBranchName,
-  [string]$prNumber)
+  [string]$prNumber,
+  [string]$commitSHA)
 
-try {  
-    # name and email are only used for merge commit, it doesn't really matter what we put in there.    
+try {
+    # name and email are only used for merge commit, it doesn't really matter what we put in there.
     git config user.name "RoslynValidation"
-    git config user.email "validation@roslyn.net"   
-    
+    git config user.email "validation@roslyn.net"
+
     if ($sourceBranchName -notlike '*-vs-deps') {
       Write-Host  "##vso[task.LogIssue type=warning;]The base branch for insertion validation is $sourceBranchName, which is not a vs-deps branch."
     }
+
+    Write-Host "Validating the PR head matches the specified commit SHA ($commitSHA)..."
+    if ($commitSHA.Length -lt 7) {
+      Write-Host "##vso[task.LogIssue type=error;]The PR Commit SHA must be at least 7 characters long."
+      exit 1
+    }
+
+    Write-Host "Getting the hash of refs/pull/$prNumber/head..."
+    $remoteRef = git ls-remote origin refs/pull/$prNumber/head
+    Write-Host ($remoteRef | Out-String)
+
+    $prHeadSHA = $remoteRef.Split()[0]
+    if (!$prHeadSHA.StartsWith($commitSHA)) {
+      Write-Host "##vso[task.LogIssue type=error;]The PR's Head SHA ($prHeadSHA) does not begin with the specified commit SHA ($commitSHA). Unreviewed changes may have been pushed to the PR."
+      exit 1
+    }
+
     Write-Host "Setting up the build for PR validation by merging refs/pull/$prNumber/merge into $sourceBranchName..."
     git pull origin refs/pull/$prNumber/merge
```

Comment:
This seems like a race condition. The remote SHA may have changed in between `ls-remote` and `pull`. Consider pulling the `$commitSHA` instead.

Reply:
I mean technically true, but would be very tight. 

Happy to defer this as a future enhancement where we also include merging the target branch in with a specific commit sha from the PR branch.

Reply:
It feels like the risk of this may be less if we fetch all the refs we want in one invocation.